### PR TITLE
Allow to rebind the old value of a mutable Ltac2 entry.

### PR DESCRIPTION
**Pull request by ppedrot**
This PR introduces an extension of the Ltac2 rebinding syntax in the form of
```
Ltac2 Set my.nifty.def as old := ...
```
that allows to access the old binding of the entry and reuse it inside its new value.

Pinging @kyoDralliam who was trying to implement that.

- [X] Added / updated test-suite
- [X] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by ppedrot** [comment]
FTR, I also want to introduce the ability to rebind such a definition at the level of terms, e.g. with a `let Set my.def := foo in ...` syntax, but this raises the question of the semantics of delayed evaluation for closures. In this setting, do they bind the current value of the constant, or the one at the time they will be evaluated?

**Comment by ppedrot** [comment]
This is ready and needs an assignee.

File: doc/changelog/05-tactic-language/11503-ltac2-rebind-with-value.rst
Comment (Zimmi48) [inline doc/changelog/05-tactic-language/11503-ltac2-rebind-with-value.rst]: That way there is a direct link to the doc:

```suggestion
  The Ltac2 rebinding command :cmd:`Ltac2 Set` has been extended with the ability to
```

**Comment by ppedrot** [comment]
@coq/ltac2-maintainers ping

File: test-suite/ltac2/rebind.v
Comment (JasonGross) [inline test-suite/ltac2/rebind.v]: I think most of these should be output tests

**Review by JasonGross — CHANGES_REQUESTED** [review]
Should I add myself to ltac2-maintainers so I can assign and approve and merge?

I don't really understand the implementation well, but I'm quite happy with the user-visible changes (though I think the new entries in the test-suite file should live in the output directory)

**Comment by JasonGross** [comment]
> FTR, I also want to introduce the ability to rebind such a definition at the level of terms, e.g. with a `let Set my.def := foo in ...` syntax, but this raises the question of the semantics of delayed evaluation for closures. In this setting, do they bind the current value of the constant, or the one at the time they will be evaluated?

My intuition is to pick whichever one has a simpler model and documentation.  In this case, my model is that there is a (global) map of names to tactics.  Writing `Ltac2 Set` updates the mapping, inlining the current value of the map wherever the `as`-bound name appears.  There is no notion of "the map before I wrote `Ltac2 Set`" which is accessible anywhere, so I think the answer to your question, on my model, is "the current value of the constant".

**Review by kyoDralliam — COMMENTED** [review]
Since the evaluation of pure terms has been extended to account for let-bindings of syntactical values, I think it would be reasonable to relax a bit the definition of syntactical values and accept these let bindings in general (not only in the special case where it comes from some code generated behind the back of the user). Otherwise we loose the invariant that tactic idents are bound to syntactical values.

**Comment by ppedrot** [comment]
@kyoDralliam done.

File: test-suite/ltac2/rebind.v
Comment (ppedrot) [reply test-suite/ltac2/rebind.v]: I can probably turn this into a proper check instead, output tests have a tendency to be fragile.

**Comment by kyoDralliam** [comment]
@ppedrot Is the mathcomp failure in CI meaningful ? (it does not seem relevant to me)

Otherwise, up to the change in the test and if possible documentation of the change to the notion of Ltac2 value, I would be ready to merge.

**Comment by ppedrot** [comment]
The CI failure is unrelated. I'll update the tests and doc asap.

**Comment by kyoDralliam** [comment]
@ppedrot Any progress on this PR ?

**Comment by ppedrot** [comment]
I can't approve my own PR, but I informally validate @kyoDralliam's changes. I'll rebase this and it should then be ready to merge.

**Comment by ppedrot** [comment]
I've assigned @JasonGross since @kyoDralliam is now a co-author.

**Review by JasonGross — CHANGES_REQUESTED** [review]
I see there is
```coq
(** Subtype check *)
Ltac2 mutable rec f x := f x.
Fail Ltac2 Set f := fun x => x.
Ltac2 mutable g x := x.

Ltac2 Set g := f.
```
in the test-suite file.  I take it this is about subtyping of the new value, but there's no other test of `Ltac2 Set` on an `Ltac2 mutable rec` definition.  If this is supposed to work, please add a test-case for it, as well as one that uses the `as lident` extension with it, to ensure that the original tactic calls itself rather than the new body.  If it's not supposed to work, please add a test checking that it fails correctly (and not with, e.g., assertion failures) and add a note to the documentation that this case does not work.  (And why is `Ltac2 Set rec qualid` not allowed?  Or is it allowed by undocumented?)

**Comment by JasonGross** [comment]
(Sorry for the very delayed change request, but I think this case is important to document/deal wth.)

**Comment by kyoDralliam** [comment]
@ppedrot Is the following behaviour intended ?
```
Ltac2 mutable x := 5.
Ltac2 y := x.
Ltac2 Eval y. (* 5 *)
Ltac2 Set x := 7.
Ltac2 Eval y. (* 7 *)
```
I would have expected that an identifier defined **without** `Set` would be a pure value and thus should not change even in presence of side effects...

**Comment by ppedrot** [comment]
@kyoDralliam this is the same problem as the semantic of closures mentioned in my second post. Currently, "mutable" Ltac2 definitions enjoy a dynamic binding semantics.

**Comment by kyoDralliam** [comment]
Another strange thing, corresponding @JasonGross request: we can `Set` a non-recursive mutable variable  to a recursive one
```
Ltac2 mutable quw := fun (n : nat) => O.
Ltac2 Set quw := fun n =>
  match n with
  | O => O
  | S n => S (S (quw n))
  end.
```

**Comment by kyoDralliam** [comment]
@ppedrot thanks for clarifying. Do you mind if I add it as a test ? It might help newcomers to ltac2 to understand what's happening.

**Comment by kyoDralliam** [comment]
I took the liberty to add the tests requested by @JasonGross.
 
I don't think we need to say anything about `Ltac2 Set rec` in the doc because re-bindings always emulate recursive definitions through their dynamic scope.
This can clearly lead to strange definitions though :
```
(* identity *)
Ltac2 mutable quu := fun (n : nat) => n.
(* doubling function *)
Ltac2 Set quu := fun n =>
               match n with
               | O => O
               | S n => S (S (quu n))
               end.

(* looks like 2*n+1 but actually 3*n+1 because of dynamic binding...*)
Ltac2 Set quu as self := fun n => S (self n).
Ltac2 Eval (quu (S (S O))). (* = 7 *)
```

**Comment by JasonGross** [comment]
> I don't think we need to say anything about `Ltac2 Set rec` in the doc because re-bindings always emulate recursive definitions through their dynamic scope.

If `Ltac2 Set rec` is valid, then the syntax in the doc should specify.  In any case, I think it should be made clear in the doc that in `Ltac2 Set foo as bar := ...`, any recursive invocations of `foo` within the body of `bar` will use the new definition of `foo`, not the old one.  (Note that this is different from how mutable recursive Ltac2 functions who access recursion via a generic Y combinator would work.)

**Comment by ppedrot** [comment]
FTR, I'll recall here some thoughts I had about the dynamic binding feature of Ltac2 some time ago. Instead of the current model where dereferencing is implicit, we could wrap mutable definitions into a type `'a cell` which stands for reader-style mutable references. In particular, there would be two primitive functions `read : 'a cell -> 'a` and `enter : ('a -> 'b) -> 'a cell -> 'a -> 'b` to interact with such cells, in addition to the toplevel `Ltac2 Set`. It would make the situation probably clearer to have this in the type system.

**Comment by kyoDralliam** [comment]
@JasonGross Sorry, I haven't been clear enough in my message. `Ltac2 Set rec` is not valid syntax and I think it's outside the scope of this PR.
There would be a few things to pounder if we were to add it: depending on the presence or the absence of `rec` the recursive occurrences of the function being redefined would obey a static or dynamic scoping discipline. Moreover the interaction of redefinitions with (implemented but undocumented) mutually recursive definitions should also be thought of.

With respect to the documentation, I have added two examples to make clearer the current behaviour of `Set` and its interaction with  recursive functions.

@ppedrot I think making explicit the dynamic nature of mutable cells in their type and usage would make a lot of sense to have saner semantics for Ltac2.
One question though: you gave `read` the type `'a cell -> 'a` but if you want to reproduce the polymorphic examples that we currently have in the doc, you would need to instanciate `'a` with a type scheme `'b. 'b -> 'b` no ? So if we want to keep things simple (no higher-rank polymorphism) `read x` and `enter f x` will probably have to be primitive constructions of the language rather than functions.

**Comment by ppedrot** [comment]
> you would need to instanciate 'a with a type scheme

@kyoDralliam you're right, I think this was one of the reasons that prevented me from starting to implement this.

**Comment by kyoDralliam** [comment]
I think it would still be worth making these `cell` explicit. I see three options:

1. We keep the current PR, mutability is implicit and "values" can actually depend on their context.
2. We introduce `'a cell` as a type, it can only contain value of monomorphic type `'a`; this rules out cells containing parametric functions or data structures but `read` and `enter` are standard functions.
3. We introduce  `('a. t) cell` as a new type-scheme, it can contain values of polymorphic type but `read` and `enter` are operators of Ltac2 with specific typing rules.

Now that I think a bit more about it, I find option 1) clashing quite a bit with the expectations provided by the other constructs of Ltac2 (the restriction of definitions to values, being explicit in binders about effects...).

I think the second case is well studied and known, probably not that difficult to implement. The third one seems much more of a challenge. Can we think of convincing examples that would require `cell`s to contain polymorphic values ?

**Comment by Zimmi48** [comment]
Do you think that this PR will be finished and merged in time for the branching on Friday?

**Comment by JasonGross** [comment]
Given the updated documentation and test-suite, I'm using to merge this if there is consensus to do so from @ppedrot and @kyoDralliam .  However, my inclination would be to delay this until just after branching, as it seems quite possible that even if this feature is merged, it'll get changed in the next release cycle, and it seems better to me if we don't have it appear in one version just to change in the next.  But, again, I'm happy to bow to consensus.

**Comment by ppedrot** [comment]
The toplevel feature is unlikely to be changed, my proposal is rather about the currently non-existing equivalent feature for inline terms, as well as tweaking the dereferencing of mutable definitions, which is already there.

**Review by JasonGross — APPROVED** [review]

**Comment by JasonGross** [comment]
CI Failures seem spurious, will merge now

**Pull request merged**
